chapter_01/final_exercise_01.md

```diff
@@ -20,27 +19,177 @@
+        So on prem it is slightly more expensive than hadoop.
+
+- [x] *Hdfs use cases*: to store large amounts of files that need to be searchable (Websites and text / pdf files), hadoop is often used in the finance industry for risk modelling (using big data that is stored in hdfs and analyzing with mapreduce and spark), can also be used like a regular file / block storage that can handle a lot of data and scale easily while being reliable, Hdfs can also be used along with analytics tools to detect fradulent activity.
+
+- [x] *Terms*:
+  - [x] *Failover*: The process of handling a node / master failure, when a leader fails, a failover event is triggered and a leader re election starts.
+  - [x] *Fencing*: Isolation of a node or shared resources when a node seems to be malfunctioning.
+  - [x] *Zkfc*:  The Zookeeper Failover Controller which is responsible for automatic failover in distributed systems like Hadoop (namenode failover) and Kafka (Controller failover)
+  - [x] *Quorum*: Quorum (in zookeeper) is the smallest amount of running servers of a distributed system in order for it to work and be available for the clients, usually the formula for a quorum is that if you have 2n + 1 nodes, n nodes can fail and the system will keep working as it worked before.
+  The quorum can also be used to elect leaders, by having a quorum of the in sync nodes that can become leaders in the case of node failure.
+
+
 1. **Q:** What is Apache Hadoop?
+1. **A:** Hadoop is a distributed data processing framework that works based on a master-slave architecture which was built with the assumption that hardware failure is the norm and allows for big data processing with easy scaling, to thousands of nodes.
+Hadoop has a few components in it that allow for optimal resource usage and job schedualing while allowing multiple jobs simountaneusly, Hadoop also includes Hdfs which is a distributed file system that saves tha
+Hadoop also contains YARN which allows for better and more efficient resource utilization in addition to allowing multiple processes to run in parrallel.
+Yarn works with a quite simple architecture, when it has the resource manager which is responsible for managing the resources of the cluster, the Node manager which Sends the status of the Node to the resource manager and monitors the resources of the Node, then it has the Application managers which are responsible for negotiating resources for an app or a process that we want to run.
+Hadoop is a rack aware system which means that hadoop knows the physical network topology (through a resolve api call through rpc) and can place the data in those placed accordingly to ensure maximum fault tolorence evenif one rack goes down
+Hadoop allows for other processing frameworks to run on it, including Spark, MapReduce, Hive, Hbase and Zepplin.
```

Comment:
Hadoop includes these "frameworks" and they are a huge part of it (Zepplin excluded).